### PR TITLE
ClassMyArrayList

### PR DESCRIPTION
**Pull request by YatKat**
My own realization of ArrayList

### CODE REVIEW

File: src/lesson5/MyArrayList.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
+package lesson5;
+
+public class MyArrayList<T> {
```
Comment (ghost) [inline src/lesson5/MyArrayList.java]: should implement an interface, see instructions to task

File: src/lesson5/MyArrayList.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        data = new Object[10];
+    }
+
+    public MyArrayList(int capacity) {
+        if (capacity < 0)
+            throw new IllegalArgumentException("Cannot construct array with negative capacity");
+        data = new Object[capacity];
+    }
+
+    public void add(T value) {
```
Comment (ghost) [inline src/lesson5/MyArrayList.java]: check return type in task

File: src/lesson5/MyArrayList.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        ensureCapacity(size + 1);
+        this.data[size] = value;
+        size++;
+    }
+
+    public void add(T value, int index) {
```
Comment (ghost) [inline src/lesson5/MyArrayList.java]: check return type in task

File: src/lesson5/MyArrayList.java [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        return size == 0;
+    }
+
+    //unchecked cast//
+    public T get(int index) {
+        checkIndex(index);
+        return (T) data[index];
+    }
+
+    public void remove(int index) {
```
Comment (ghost) [inline src/lesson5/MyArrayList.java]: check return type in task

File: src/lesson5/MyArrayList.java [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        if (data.length >= capacity) return;
+        Object[] newData = new Object[data.length * 2];
+        for (int i = 0; i < data.length; i++) {
+            newData[i] = data[i];
+        }
+        data = newData;
+    }
+
+    public void checkIndex(int index) {
+        if (index > size || index < 0) {
```
Comment (ghost) [inline src/lesson5/MyArrayList.java]: check how it works on a different indexes

File: src/lesson5/MyArrayList.java [hunk 6]
```diff
@@ -1,6 +1,11 @@
 package lesson5;
 
-public class MyArrayList<T> {
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+public class MyArrayList<T> implements List{
```
Comment (boroda4436) [inline src/lesson5/MyArrayList.java]: Implement List  <T>

File: src/lesson5/Main.java [hunk 7]
```diff
@@ -0,0 +1,23 @@
...
+
+        System.out.println("Size " + list.size());
+        System.out.println("Contains Lis " + list.contains("Lis"));
+        System.out.println("The element " + list.remove(2) + " removed");
+        System.out.println("Size " + list.size());
+        list.add("Zhyrik", 1);
+        System.out.println(list.toString());
+    }
+
+
```
Comment (boroda4436) [inline src/lesson5/Main.java]: Empty lines

File: src/lesson5/Main.java [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+import java.util.List;
+
+public class Main {
+    public static void main(String[] args) {
+        MyArrayList<String> list = new MyArrayList<>();
+        list.add("Barsik");
+        list.add("Murzik");
+        list.add("Ryzhyk");
+        list.add("Lis");
+
```
Comment (boroda4436) [inline src/lesson5/Main.java:10]: Add more test cases for methods add (index, value)

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/lesson5/MyArrayList.java [hunk 9]
```diff
@@ -0,0 +1,197 @@
+package lesson5;
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+public class MyArrayList<T> implements List {
```
Comment (boroda4436) [inline src/lesson5/MyArrayList.java]: Implements parametrized List<T> required!

File: src/lesson5/MyArrayList.java [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+    }
+
+    @Override
+    public Object set(int index, Object element) {
+        return null;
+    }
+
+    @Override
+    public void add(int index, Object element) {
+
```
Comment (boroda4436) [inline src/lesson5/MyArrayList.java]: Need to implement this one for backword compatability

File: src/lesson5/MyArrayList.java (hunk 3 above)
Comment (boroda4436) [reply src/lesson5/MyArrayList.java]: This method does not exist in List interface

**Review by boroda4436 — CHANGES_REQUESTED** [review]

File: src/lesson5/MyArrayList.java [hunk 11]
```diff
@@ -0,0 +1,193 @@
...
+    private void ensureCapacity(int capacity) {
+        if (data.length >= capacity) return;
+        Object[] newData = new Object[data.length * 2];
+        for (int i = 0; i < data.length; i++) {
+            newData[i] = data[i];
+        }
+        data = newData;
+    }
+
+    public void checkIndex(int index) {
```
Comment (boroda4436) [inline src/lesson5/MyArrayList.java]: Should this method be public or private?

File: src/lesson5/MyArrayList.java [hunk 12]
```diff
@@ -0,0 +1,193 @@
...
+        return (T) data[index];
+    }
+
+    @Override
+    public Object set(int index, Object element) {
+        return null;
+    }
+
+    @Override
+        public void add(int index, T element) {
```
Comment (boroda4436) [inline src/lesson5/MyArrayList.java]: Apply code autoformating

File: src/lesson5/Main.java [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+        System.out.println("Size " + list.size());
+        System.out.println("Contains Lis " + list.contains("Lis"));
+        System.out.println("The element " + list.remove(2) + " removed");
+        System.out.println("Size " + list.size());
+        list.add("Zhyrik", 1);
```
Comment (boroda4436) [inline src/lesson5/Main.java]: does this method still exist? No compilation error here?

**Review by boroda4436 — CHANGES_REQUESTED** [review]

**Review by boroda4436 — APPROVED** [review]
LGTM

**Pull request merged**